### PR TITLE
Add export and project duplication

### PR DESCRIPTION
**Pull request by hardillb**
Contains a Vue component for selecting which parts of a Project should be exported. This will be reused for Project duplicate workflow.

Work in progress towards https://github.com/flowforge/flowforge/issues/270 and #271

### CODE REVIEW

**Comment by hardillb** [comment]
@joepavitt can you review the vue in this please. I'm sure it's ugly as hell but it does work...

**Comment by hardillb** [comment]
I'm starting to think I should disable the export (to JSON blob as that is covered in a different story not in the 0.5 release plan and there is nothing you can do with it yet)

File: frontend/src/pages/project/Settings/Danger.vue [hunk 1]
```diff
@@ -38,6 +68,38 @@ export default {
...
+        },
+        showExportToProjectDialog () {
+            this.$refs.exportToProjectDialog.show(this.project)
+        },
+        exportProject (parts) {
+            // call projectApi to generate zipped json
+            projectApi.exportProject(this.project.id, parts)
+        },
+        duplicateProject (parts) {
+            console.log('copy project', parts)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:85]:
```suggestion
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 2]
```diff
@@ -38,6 +68,38 @@ export default {
...
+            projectApi.create(parts).then(result => {
+                console.log(result)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:87]:
```suggestion
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 3]
```diff
@@ -38,6 +68,38 @@ export default {
...
+                console.log(result.id)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:88]:
```suggestion
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 4]
```diff
@@ -38,6 +68,38 @@ export default {
...
+                console.log(result.name)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:89]:
```suggestion
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 5]
```diff
@@ -38,6 +68,38 @@ export default {
...
+                this.$router.push({ name: 'Project', params: { id: result.id } })
+            }).catch(err => {
+                console.log(err)
+            })
+        },
+        exportToProject (parts) {
+            console.log('Ben')
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:96]:
```suggestion
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 6]
```diff
@@ -38,6 +68,38 @@ export default {
...
+            console.log(parts)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/Danger.vue:97]:
```suggestion
```

File: frontend/src/pages/project/Settings/dialogs/ExportToProjectDialog.vue [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                                    <p class="text-sm text-gray-500">
+                                        Select project components to copy to project:
+                                    </p>
+                                </div>
+                                <div>
+                                    <FormRow :error="errors.target" v-model="input.target" :options="projects">
+                                        <template v-slot:default>Target Project</template>
+                                    </FormRow>
+                                </div>
+<!--
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/dialogs/ExportToProjectDialog.vue]: Remove commented out code

File: frontend/src/pages/project/Settings/dialogs/ExportToProjectDialog.vue [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                }
+                isOpen.value = true
+                this.input.stack = this.project.stack.id
+                this.input.template = this.project.template.id
+                this.input.team = this.project.team.id
+                const projectList = await teamApi.getTeamProjects(this.project.team.id)
+                this.projects = []
+                for (let i = 0; i < projectList.count; i++) {
+                    if (this.project.id !== projectList.projects[i].id) {
+                        console.log(projectList.projects[i].id, projectList.projects[i].name)
```
Comment (joepavitt) [inline frontend/src/pages/project/Settings/dialogs/ExportToProjectDialog.vue:174]:
```suggestion
```

File: frontend/src/pages/project/components/ExportProjectComponents.vue [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        </FormRow>
+        <FormRow class="ml-9" v-if="parts.envVars" type="radio" v-model="parts.envVarsKo" :value="false">
+            Keys and Values
+            <!-- <template v-slot:description>Keys and Values</template> -->
+        </FormRow>
+        <FormRow class="ml-9" v-if="parts.envVars" type="radio" v-model="parts.envVarsKo" :value="true">
+            Keys only
+            <!-- <template v-slot:description>Keys only</template> -->
+        </FormRow>
+        <!-- <FormRow type="checkbox" v-model="parts.settings">Settings</FormRow> -->
```
Comment (joepavitt) [inline frontend/src/pages/project/components/ExportProjectComponents.vue:27]: Remove commented out code

File: frontend/src/pages/project/components/ExportProjectComponents.vue [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            Environment Variables
+            <template #description>Export Environment Variables</template>
+        </FormRow>
+        <FormRow class="ml-9" v-if="parts.envVars" type="radio" v-model="parts.envVarsKo" :value="false">
+            Keys and Values
+            <!-- <template v-slot:description>Keys and Values</template> -->
+        </FormRow>
+        <FormRow class="ml-9" v-if="parts.envVars" type="radio" v-model="parts.envVarsKo" :value="true">
+            Keys only
+            <!-- <template v-slot:description>Keys only</template> -->
```
Comment (joepavitt) [inline frontend/src/pages/project/components/ExportProjectComponents.vue:25]: Remove commented out code

**Review by joepavitt — CHANGES_REQUESTED** [review]
I will likely modify a few of the dialogs shortly, once I proceed with the update to `forge-ui-components` dialog boxes across the platform.

Mostly looking good, but a lot of commented code and console logs that need clearing.

**Review by joepavitt — APPROVED** [review]

**Comment by hardillb** [comment]
Still trying to work out how to clean up the overflow here:

Screenshot from 2022-05-05 16-06-21

And how to line up:

Screenshot from 2022-05-05 15-41-06

**Comment by joepavitt** [comment]
> Still trying to work out how to clean up the overflow here

This is an issue with the existing FormRow. I'm honestly okay with pushing as is, with the knowledge that this will be cleaned by as part of the [Dialog Box rework](https://github.com/flowforge/flowforge/issues/520) I'm due to do in 0.5

**Comment by knolleary** [comment]
#273 (export to zip) is not in plan for 0.5. We haven't done any design work around how the zip is structured. Given this PR implements something, can we get a description of what the zip contents is?

**Comment by hardillb** [comment]
The JSON output is as follows:

```json
{
   "flows": [],
   "credentials": { "$": "b5db5a7f0aca0467b928136ff4c01e0flfGwBx7CcOrJHef+xp5StVrny+QwoLhn48aLck+gtIGXp012ujlkmNaTW2vJ0I8gyiAY06iZpbDb" },
  "envVars": {
    "FOO": "BAR",
    "TESTING": "ONE"
  },
  "nodes": {
    "node-red": {...},
    "node-red-node-base64": { ... }
  }
}
```

Where `flows` is the array of tabs/nodes and the `nodes` is the nodes section from the settings object. This was a bare minimum I thought we needed. Open to adding something about stacks and templates, but would need to decide what from each.

**Comment by hardillb** [comment]
We can always remove the UI component for the export to JSON option if needed which will hide this option until we can iterate on it.

**Comment by knolleary** [comment]
I would prefer we didn't expose this option today.

It will come in useful under the covers for the Device work - but that will also help us validate the format and the contents.

**Comment by knolleary** [comment]
I find the reworked UI (buttons above text) to be top heavy and imbalanced. I think the buttons are too big.

I would prefer to get back to the original intent of having the button in a right-hand column - but know full well why that's a little awkward without a bigger refactor of FormRow.

@joepavitt would setting `size="small"` be inoffensive?

**Comment by joepavitt** [comment]
> @joepavitt would setting `size="small"` be inoffensive?

No objection at all from me

File: frontend/src/pages/project/Settings/Danger.vue [hunk 11]
```diff
@@ -1,23 +1,51 @@
...
+                project's current state.</div>
+        </div> -->
+
+        <div>
+            <div>
+                <ff-button kind="secondary" @click="showDuplicateProjectDialog()">Duplicate Project</ff-button>
+                <DuplicateProjectDialog @duplicateProject="duplicateProject" ref="duplicateProjectDialog"/>
+            </div>
+            <div class="max-w-sm pt-2 pl-5">Allows you to use a snapshot of the
+                project's current state to create a new project.</div>
```
Comment (knolleary) [inline frontend/src/pages/project/Settings/Danger.vue:31]:
```suggestion
            <div class="max-w-sm pt-2 pl-5">Create a copy of this project.</div>
```

File: frontend/src/pages/project/Settings/Danger.vue [hunk 12]
```diff
@@ -1,23 +1,51 @@
...
+        </div>
+
+        <div>
+            <div>
+                <ff-button kind="secondary" @click="showExportToProjectDialog()">Export State</ff-button>
+                <ExportToProjectDialog @exportToProject="exportToProject" ref="exportToProjectDialog"/>
+            </div>
+            <div class="max-w-sm pt-2 pl-5">Copy a project's state to another existing project.</div>
```
Comment (knolleary) [inline frontend/src/pages/project/Settings/Danger.vue:39]:
```suggestion
                <ff-button kind="secondary" @click="showExportToProjectDialog()">Copy to existing project</ff-button>
                <ExportToProjectDialog @exportToProject="exportToProject" ref="exportToProjectDialog"/>
            </div>
            <div class="max-w-sm pt-2 pl-5">Copy the project's state to an existing project.</div>
```

**Review by knolleary — COMMENTED** [review]
Some suggested tweaks to the labels/descriptions

**Comment by knolleary** [comment]
Just tried the 'Duplicate project' - kept all the default options - and got this error in the runtime:
```
 req: {
      "method": "POST",
      "url": "/api/v1/projects",
      "hostname": "localhost:3000",
      "remoteAddress": "127.0.0.1",
      "remotePort": 55949
    }
    res: {
      "statusCode": 500
    }
    reqId: "req-s"
    err: {
      "type": "TypeError",
      "message": "Cannot read properties of null (reading 'flow')",
      "stack":
          TypeError: Cannot read properties of null (reading 'flow')
              at Object.<anonymous> (/Users/nol/code/flowforge/flowforge/forge/routes/api/project.js:208:39)
    }
```

Note - this is a NR Project that doesn't actually have any flows as I'm using the stub driver.

---

For the Duplicate project UI, I wonder if we'd be better off taking the user to the existing 'create project' page - but with an added 'copy content from...' section - rather than doing this all via a small dialog box.

Having just one view that handles creating projects would remove duplication in UI code. It would also, for example, ensure the 'agree to extra charges' checkbox is there when billing is enabled.

I am happy to tackle that rework if we think its worthwhile.

**Comment by hardillb** [comment]
OK, I will ladd tests for missing flows/creds/settings to the project endpoints (e.g when using the stub driver, but also for a project that hasn't deployed yet)

No problem with moving to the create view, but how will it handle the choices of what to duplicate? Or will you pull in the new component into the create vue?

**Comment by knolleary** [comment]
> Or will you pull in the new component into the create vue?

Yes - reduce, reuse, recycle.

**Comment by hardillb** [comment]
Sounds good, just working out what to with the export to project, if the source project doesn't have the things it should be exporting. Should it just replace the existing with empty values?

**Pull request closed without merging**